manual/node_pattern.md

```diff
@@ -0,0 +1,292 @@
+methods.
+
+The code below belongs to [Style/ArrayJoin](http://www.rubydoc.info/github/bbatsov/rubocop/Rubocop/Cop/Style/ArrayJoin)
+cop and it's in favor of `Array#join` over `Array#*`. Then it tries to find
+code like `%w(one two three) * ", "` and suggest to use `#join` instead.
+
+It checks if the code is an array of strings and call `*` in the end:
+
+```ruby
+def on_send(node)
+  receiver_node, method_name, *arg_nodes = *node
+  return unless receiver_node && receiver_node.array_type? &&
+    method_name == :* && arg_nodes.first.str_type?
+
+  add_offense(node, location: :selector)
+end
+```
+
+This code was replaced in the cop defining a new matcher that does the same as the code above:
+
+```ruby
+def_node_matcher :join_candidate?, '(send $array :* $str)'
+```
+
+And the `on_send` method is simplified to a simple method usage:
+
+```ruby
+def on_send(node)
+  join_candidate?(node) { add_offense(node, location: :selector) }
+end
+```
+
+## Simple match
+
+A simple integer like `1` represents `(int 1)` in the AST.
+
+```
+$ ruby-parse -e '1'
+(int 1)
+```
+
+- `int` will match exactly the node, looking only the node type.
+- `(int 1)` will match precisely the node
+
+## `_` for any single node
+
+`_` will check if there's something present in the specific position, no matter the
+value:
+
+- `(int _)` will match any number
+- `(int _ _)` will not match because `int` types have just one child that
+  contains the value.
+
+
+## `...` for several subsequent nodes
+
+Where `_` matches a single node, `...` eagerly matches one or more subsequent nodes.
+
+It's useful when you want to check some variable internal nodes but with a
+final with the same results. For example, let's use a classic `sum(1,2)`.
+
+We can also have `sum(1,2,3,n)` and the arguments can vary. The objective is
+match all. So, let's check how it looks like in the AST:
+
+```
+$ ruby-parse -e 'sum(1,2)'
+(send nil :sum
+  (int 1)
+  (int 2))
+```
+
+Or with more children:
+
+```
+$ ruby-parse -e 'sum(1,2,3,n)'
+(send nil :sum
+  (int 1)
+  (int 2)
+  (int 3)
+  (send nil :n))
+```
+
+The first case can be addressed with an expression like:
+
+```
+(send nil? :sum ...)
+```
+
+## `{}` for any `<expression>`
+
+Lets make it a bit more complex and introduce floats:
+
+```
+$ ruby-parse -e '1'
+(int 1)
+$ ruby-parse -e '1.0'
+(float 1.0)
+```
+
+- `({int float} _)` - int or float types, no matter the value
+
+## `$` for captures
+
+You can capture elements or nodes along with your search, prefixing the expression
+with `$`. For example, in a tuple like `(int 1)`, you can capture the value using `(int $_)`.
+
+You can also capture multiple things like:
+
+```
+(${int float} $_)
+```
+
+It will return an array with `[:int, 1]` for a tuple like (int 1).
+The tuple can be entirely captured using the `$` before the open parens:
+
+```
+$({int float} _)
+```
+
+Or remove the parens and match directly from node head:
+
+```
+${int float}
```

Comment:
This is a consequence that `(int _)` works in the same way as `int`, and `$({_ _})` as `${(...) (...)}`, but it's hard to comprehend it just at this point of tutorial.